core/string/src/display.rs

```diff
@@ -0,0 +1,25 @@
+//! Display implementations for [`crate::JsString`].
+use crate::{JsStr, JsStrVariant};
+use std::fmt;
+use std::fmt::Write;
+
+/// Display implementation for `JsString` that escapes unicode characters.
+#[derive(Debug)]
+pub struct JsStringDisplayEscaped<'a> {
+    pub(crate) inner: JsStr<'a>,
+}
+
+impl fmt::Display for JsStringDisplayEscaped<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self.inner.variant() {
+            // SAFETY: `JsStrVariant::Latin1` is always valid utf8, so no need to check.
+            JsStrVariant::Latin1(v) => unsafe { std::str::from_utf8_unchecked(v) }.fmt(f),
```

Comment:
Actually, this is not correct! ascii (uses 7-bits) is always valid utf8, but not latin1 (uses full 8-bits).

Instead the implementation could use the `iter` method and would eliminate the match as well.

Reply:
I used the variant name for a reason; `JsStrVariant::Latin1` is improper naming but should be its own fix and not part of this PR.

Actually double checked the usage of Latin1 in JsString and co, and I think it's potentially unsafe. I'll file a bug tomorrow.

> Instead the implementation could use the code_points method and would eliminate the match as well.

Reply:
I did a thing; https://github.com/boa-dev/boa/pull/3958. I can fix it but it's not in scope for this PR.